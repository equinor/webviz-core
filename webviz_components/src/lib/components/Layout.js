import React, {Component} from 'react';
import {BrowserRouter, Route, Switch} from 'react-router-dom';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Menu from '../private-components/Menu';
import Banner from '../private-components/Banner';
import '../resources/dash.css';

const PageWrapper = styled.div`
    display: flex;
    flex-direction: column;
    width: 100%;
`;

const Wrapper = styled.div`
    display: flex;
    height: 100%;
    @media (max-width: 1200px) {
        flex-direction: column;
    }
`;

class Layout extends Component {
    render() {
        const {children, banner, basepath} = this.props;
        const [frontPage, ...subPages] = Array.isArray(children)
            ? children
            : [children];
        const subPageProps = subPages.map(({props}) => {
            // If dashprivate_layout, get props from there
            // Else, get props from component itself
            const props2 =
                props._dashprivate_layout && props._dashprivate_layout.props
                    ? props._dashprivate_layout.props
                    : props;

            if (props2 && !props2.title && props2.children) {
                return props2.children.props;
            }

            return props2;
        });
        const pages = subPages.map((page) => {
            // If dashprivate_layout, get props from there
            // Else, get props from component itself
            const pageProps =
                page.props._dashprivate_layout &&
                page.props._dashprivate_layout.props
                    ? page.props._dashprivate_layout.props
                    : page.props;

            <Route
                key={pageProps.id}
                path={`${basepath}/${pageProps.id}`}
                render={() => <PageWrapper>{page}</PageWrapper>}
            />;
        });
        return (
            <BrowserRouter>
                <Wrapper id="layout">
                    <Menu basePath={basepath} subPages={subPageProps} />
                    <Switch>
                        <Route
                            exact
                            path={`${basepath}/`}
                            render={() => (
                                <PageWrapper>
                                    {banner && <Banner {...banner} />}
                                    {frontPage}
                                </PageWrapper>
                            )}
                        />
                        {pages}
                    </Switch>
                </Wrapper>
            </BrowserRouter>
        );
    }
}

Layout.defaultProps = {
    children: null,
    banner: null,
    basepath: '',
};

Layout.propTypes = {
    /**
     * The content of the layout. Ideally a list of pages or a single page
     */
    children: PropTypes.node,
    /**
     * The banner is optional and is only the front page.
     * If you wish to have a banner provide dict containing a color or a url
     * and a title
     */
    banner: PropTypes.shape({
        url: PropTypes.string,
        color: PropTypes.string,
        title: PropTypes.string,
    }),
    basepath: PropTypes.string,
};

export default Layout;
