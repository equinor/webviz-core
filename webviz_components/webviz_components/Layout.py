# AUTO GENERATED FILE - DO NOT EDIT

from dash.development.base_component import Component, _explicitize_args


class Layout(Component):
    """A Layout component.


Keyword arguments:
- children (a list of or a singular dash component, string or number; optional): The content of the layout. Ideally a list of pages or a single page
- banner (optional): The banner is optional and is only the front page.
If you wish to have a banner provide dict containing a color or a url
and a title. banner has the following type: dict containing keys 'url', 'color', 'title'.
Those keys have the following types:
  - url (string; optional)
  - color (string; optional)
  - title (string; optional)
- basepath (string; optional)"""
    @_explicitize_args
    def __init__(self, children=None, banner=Component.UNDEFINED, basepath=Component.UNDEFINED, **kwargs):
        self._prop_names = ['children', 'banner', 'basepath']
        self._type = 'Layout'
        self._namespace = 'webviz_components'
        self._valid_wildcard_attributes =            []
        self.available_properties = ['children', 'banner', 'basepath']
        self.available_wildcard_properties =            []

        _explicit_args = kwargs.pop('_explicit_args')
        _locals = locals()
        _locals.update(kwargs)  # For wildcard attrs
        args = {k: _locals[k] for k in _explicit_args if k != 'children'}

        for k in []:
            if k not in args:
                raise TypeError(
                    'Required argument `' + k + '` was not specified.')
        super(Layout, self).__init__(children=children, **args)
