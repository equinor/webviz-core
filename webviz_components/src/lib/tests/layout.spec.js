import React from 'react';
import renderer from 'react-test-renderer';
import Layout from '../components/Layout';

describe('Layout', () => {
    it('should render Layout with one child element', () => {
        const tree = renderer.create(
            <Layout
                _dashprivate_layout={{
                    props: {
                        children: [
                            <div id="page" title="page">
                                Page
                            </div>,
                        ],
                    },
                }}
            />
        );
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('should render Layout with multiple child elements', () => {
        const tree = renderer.create(
            <Layout
                _dashprivate_layout={{
                    props: {
                        children: [
                            <div id="frontpage" title="Frontpage">
                                Frontpage
                            </div>,
                            <div id="subpage 1" title="Subpage 1">
                                Subpage 1
                            </div>,
                            <div id="subpage 2" title="Subpage 2">
                                Subpage 2
                            </div>,
                        ],
                    },
                }}
            />
        );
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('should render Layout with banner', () => {
        const tree = renderer.create(
            <Layout
                _dashprivate_layout={{
                    props: {
                        banner: {
                            color: '#333333',
                            title: 'Banner',
                        },
                        children: [
                            <div id="page" title="Page">
                                Page
                            </div>,
                        ],
                    },
                }}
            />
        );
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('should not compare props before updates by default', () => {
        jest.mock('deep-equal', () => jest.fn(() => true));

        const equal = require('deep-equal'); // eslint-disable-line global-require
        renderer.create(
            <Layout
                _dashprivate_layout={{
                    props: {
                        children: [
                            <div id="page" title="Page">
                                Page
                            </div>,
                        ],
                    },
                }}
            />
        );
        expect(equal).toHaveBeenCalledTimes(0);
        equal.mockRestore();
    });

    it('should compare props before updates if strictUpdates is true', () => {
        jest.mock('deep-equal', () => jest.fn(() => true));

        const equal = require('deep-equal'); // eslint-disable-line global-require
        renderer.create(
            <Layout
                strictUpdate
                _dashprivate_layout={{
                    props: {
                        children: [
                            <div id="page" title="Page">
                                Page
                            </div>,
                        ],
                    },
                }}
            />
        );
        expect(equal).toHaveBeenCalledTimes(0);
        equal.mockRestore();
    });
});
