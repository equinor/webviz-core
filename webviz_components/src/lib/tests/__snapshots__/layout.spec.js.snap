// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout should render Layout with banner 1`] = `
.c5 {
  color: var(--menuLinkColor);
  height: 50px;
  width: 50px;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c5:hover {
  color: var(--menuLinkHoverColor);
}

.c1 {
  background: var(--menuBackground);
}

.c2 {
  background: var(--menuBackground);
  min-width: 250px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 30px;
  font-familiy: system-ui;
}

.c6 {
  margin: 12px 20px 20px 18px;
  margin: 20px 30px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c10 {
  background-color: #333333;
  height: 534px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c11 {
  color: #ffffff;
  font-size: 48px;
  font-family: var(--bannerFontFamily);
  font-weight: var(--bannerFontWeight);
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
}

@media (min-width:1201px) {
  .c1 {
    min-height: -webkit-fill-available;
    min-height: -moz-available;
    min-height: fill-available;
  }
}

@media (max-width:1200px) {
  .c8 {
    display: none;
  }
}

@media (min-width:1201px) {
  .c8 {
    height: 100%;
  }
}

@media (max-width:1200px) {
  .c7 {
    text-align: center;
    margin-left: -50px;
  }
}

@media (max-width:1200px) {
  .c6 {
    height: 50px;
    margin: 15px;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    width: 100%;
  }
}

@media (min-width:1201px) {
  .c4 {
    display: none;
  }
}

@media (max-width:1200px) {
  .c10 {
    height: 380px;
  }
}

@media (max-width:1200px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

<div
  className="c0"
  id="layout"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <button
            className="c5"
            onClick={[Function]}
          >
            <svg
              fill="None"
              height="18px"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="4"
              width="22px"
            >
              <path
                d="M2,2 L20,2 M2,9 L20,9 M2,16 L20,16"
              />
            </svg>
          </button>
        </div>
        <div
          className="c6"
        >
          <a
            className="c7"
            href="/"
            onClick={[Function]}
          >
            <div
              id="logo"
            />
          </a>
        </div>
      </div>
      <div
        className="c8"
      />
    </div>
  </div>
  <div
    className="c9"
  >
    <div
      className="c10"
      color="#333333"
      role="banner"
    >
      <p
        className="c11"
      >
        Banner
      </p>
    </div>
    <div
      id="page"
      title="Page"
    >
      Page
    </div>
  </div>
</div>
`;

exports[`Layout should render Layout with multiple child elements 1`] = `
.c5 {
  color: var(--menuLinkColor);
  height: 50px;
  width: 50px;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c5:hover {
  color: var(--menuLinkHoverColor);
}

.c1 {
  background: var(--menuBackground);
}

.c2 {
  background: var(--menuBackground);
  min-width: 250px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 30px;
  font-familiy: system-ui;
}

.c6 {
  margin: 12px 20px 20px 18px;
  margin: 20px 30px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c9 {
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 20px 30px;
  font-size: 16px;
  font-family: var(--menuLinkFont);
  font-weight: var(--menuLinkFontWeight);
  font-style: var(--menuLinkFontStyle);
  border-bottom: 1px solid white;
  color: var(--menuLinkColor);
}

.c9:hover {
  background: var(--menuLinkBackgroundHover);
  color: var(--menuLinkHoverColor);
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
}

@media (min-width:1201px) {
  .c1 {
    min-height: -webkit-fill-available;
    min-height: -moz-available;
    min-height: fill-available;
  }
}

@media (max-width:1200px) {
  .c8 {
    display: none;
  }
}

@media (min-width:1201px) {
  .c8 {
    height: 100%;
  }
}

@media (max-width:1200px) {
  .c7 {
    text-align: center;
    margin-left: -50px;
  }
}

@media (max-width:1200px) {
  .c6 {
    height: 50px;
    margin: 15px;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    width: 100%;
  }
}

@media (min-width:1201px) {
  .c4 {
    display: none;
  }
}

@media (max-width:1200px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

<div
  className="c0"
  id="layout"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <button
            className="c5"
            onClick={[Function]}
          >
            <svg
              fill="None"
              height="18px"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="4"
              width="22px"
            >
              <path
                d="M2,2 L20,2 M2,9 L20,9 M2,16 L20,16"
              />
            </svg>
          </button>
        </div>
        <div
          className="c6"
        >
          <a
            className="c7"
            href="/"
            onClick={[Function]}
          >
            <div
              id="logo"
            />
          </a>
        </div>
      </div>
      <div
        className="c8"
      >
        <a
          className="c9"
          href="/subpage 1"
          onClick={[Function]}
          selected={null}
        >
          Subpage 1
        </a>
        <a
          className="c9"
          href="/subpage 2"
          onClick={[Function]}
          selected={null}
        >
          Subpage 2
        </a>
      </div>
    </div>
  </div>
  <div
    className="c10"
  >
    <div
      id="frontpage"
      title="Frontpage"
    >
      Frontpage
    </div>
  </div>
</div>
`;

exports[`Layout should render Layout with one child element 1`] = `
.c5 {
  color: var(--menuLinkColor);
  height: 50px;
  width: 50px;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c5:hover {
  color: var(--menuLinkHoverColor);
}

.c1 {
  background: var(--menuBackground);
}

.c2 {
  background: var(--menuBackground);
  min-width: 250px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 30px;
  font-familiy: system-ui;
}

.c6 {
  margin: 12px 20px 20px 18px;
  margin: 20px 30px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
}

@media (min-width:1201px) {
  .c1 {
    min-height: -webkit-fill-available;
    min-height: -moz-available;
    min-height: fill-available;
  }
}

@media (max-width:1200px) {
  .c8 {
    display: none;
  }
}

@media (min-width:1201px) {
  .c8 {
    height: 100%;
  }
}

@media (max-width:1200px) {
  .c7 {
    text-align: center;
    margin-left: -50px;
  }
}

@media (max-width:1200px) {
  .c6 {
    height: 50px;
    margin: 15px;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    width: 100%;
  }
}

@media (min-width:1201px) {
  .c4 {
    display: none;
  }
}

@media (max-width:1200px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

<div
  className="c0"
  id="layout"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <button
            className="c5"
            onClick={[Function]}
          >
            <svg
              fill="None"
              height="18px"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="4"
              width="22px"
            >
              <path
                d="M2,2 L20,2 M2,9 L20,9 M2,16 L20,16"
              />
            </svg>
          </button>
        </div>
        <div
          className="c6"
        >
          <a
            className="c7"
            href="/"
            onClick={[Function]}
          >
            <div
              id="logo"
            />
          </a>
        </div>
      </div>
      <div
        className="c8"
      />
    </div>
  </div>
  <div
    className="c9"
  >
    <div
      id="page"
      title="page"
    >
      Page
    </div>
  </div>
</div>
`;
